### DISCUSSION TITLE
Arithmatex and tight lists

### DISCUSSION THREAD

**Discussion by bandokihiro** [facelessuser/pymdown-extensions#1957] [Q&A] [answered]
Hi!

The syntax of Arithmatex for block equations requires that they are preceded and followed by an empty line. Is it possible to use block equations within a tight list? I believe this requires the absence of empty lines and, for a reason that is not completely clear to me, disrupts my list. The item containing the block equation looks fine but the following item requires an extra empty line which turns the list into a loose one. I would like to construct a tight list like follows
```
-  first item
-  second item followed by a block equation (thus an empty line)

    $$x=1$$

    I can continue the second item here with this indentation

- I need an empty line above for this to be a new item (not sure why) introducing extra vertical space
```

Thanks!

**Comment by facelessuser** [comment, marked as answer]
I'm not sure I understand what the question is. Markdown uses blocks. Blocks are separated by new lines.

How do I have two separate paragraphs if I push them together?

```markdown
- I am a paragraph.
    Am I a new paragraph
    or am I a continuation?
```

Tight lists vs loose lists are often based on content. If I have a single paragraph, we can get away with not wrapping the content in a `<p>` tag and just allow it to be wrapped by the `<li>` tag. This is a tight list.

```markdown
- I am a paragraph
```

```html
<li>I am a paragraph</p>
```

If we have two paragraphs, we must wrap them so they are displayed properly.

```markdown
- I am a paragraph

    Am I a new paragraph
    or am I a continuation?
```

```
<li>
<p>I am a paragraph.</p>
<p>Am I a new paragraph
or am I a continuation?</p>
</li>
```

Even if I let you smoosh the paragraphs and Arithmatex block together, it isn't going to render as a tight list, it will render as a loose list with the paragraphs before and after wrapped in `<p>` tags as we have 3 separate blocks.

Or are you complaining about the list not being "tight" in the source, and you don't care about the rendering? In that case, we're talking about something different. Is this what you want?

```
-  first item
-  second item followed by a block equation (thus an empty line)
    $$x=1$$
    I can continue the second item here with this indentation
- I need an empty line above for this to be a new item (not sure why) introducing extra vertical space
```

Each block above is a new paragraph, why are we not separating them as such? I cannot display two paragraphs normally without a new line, so why should math blocks be any different?

This is this way because Python Markdown encourages this approach. To get what your asking, you have to do more work to get that.

Could we force the above to be handled as separate blocks? Technically, we could. It is a bit more complicated, and I see little gain to test every paragraph looking for a potential math block and then trying to parse it as multiple blocks, but we could technically attempt to start splitting paragraphs if we find block math syntax. I think it is far more explicit and clear to have the blocks separated as separate blocks as that is literally how they are represented in HTML, and by default, that is what works best in Python Markdown, but we could potentially force it to figure this out for us. I would consider this a low priority request, but it is something we could consider.

If you absolutely need to have math parsed without new lines, you could registered the provided Math SuperFences blocks as mentioned here: https://facelessuser.github.io/pymdown-extensions/extensions/arithmatex/#alternative-math-blocks.

Then you can do something like this:

````
-  first item
-  second item followed by a block equation (thus an empty line)
   ```math
   x=1
   ```
   I can continue the second item here with this indentation
- I need an empty line above for this to be a new item (not sure why) introducing extra vertical space
````

It still renders as loose, but your source will be "tight".  The reason this works is because code blocks are parsed very differently.

**Comment by bandokihiro** [comment]
Thank you for the detailed response. It makes sense for the item containing the equation to be made of three separate blocks. With the following code
```
-  first item
-  second item followed by a block equation (thus an empty line)

    $$x=1$$

    I can continue the second item here with this indentation

- third item
- fourth item
- fifth item
```
I get an extra space between the "third item" and "fourth item". But it doesn't work if I get rid of the new line in the code before "third item".
![Screenshot from 2023-02-21 08-17-18](https://user-images.githubusercontent.com/39278362/220400916-d091f436-5e24-4fe5-a82b-eb1e9582a528.png)

**Reply by facelessuser** [reply]
Yes, but this isn't an Arithmatex issue, this is how Python Markdown handles lists. Maybe this alternative list plugin is more to your liking? I haven't used it though: https://github.com/radude/mdx_truly_sane_lists.

Also, this could just be a styling issue. When paragraphs are wrapped in `<p>` there may be a margin or padding added to the end of `<p>`.  You could use CSS to not do this on the last `<p>` of and `<li>` if the `<li>` is not the last child of the `<ul>`.

**Reply by bandokihiro** [reply]
There seems to be some compatibility issues with mdx_truly_sane_lists. For now, I think I am going to live with the extra space, thanks for your help.
